src/main/java/org/disq_bio/disq/HtsjdkReadsRddStorage.java

```diff
@@ -0,0 +1,280 @@
+   * @return the current {@link HtsjdkReadsRddStorage}
+   */
+  public HtsjdkReadsRddStorage splitSize(int splitSize) {
+    this.splitSize = splitSize;
+    return this;
+  }
+
+  /**
+   * @param validationStringency the validation stringency for reading
+   * @return the current {@link HtsjdkReadsRddStorage}
+   */
+  public HtsjdkReadsRddStorage validationStringency(ValidationStringency validationStringency) {
+    this.validationStringency = validationStringency;
+    return this;
+  }
+
+  /**
+   * @param useNio whether to use NIO or the Hadoop filesystem (default) for file operations
+   * @return the current {@link HtsjdkReadsRddStorage}
+   */
+  public HtsjdkReadsRddStorage useNio(boolean useNio) {
+    this.useNio = useNio;
+    return this;
+  }
+
+  /**
+   * @param referenceSourcePath path to the reference; only required when reading CRAM.
+   * @return the current {@link HtsjdkReadsRddStorage}
+   */
+  public HtsjdkReadsRddStorage referenceSourcePath(String referenceSourcePath) {
+    this.referenceSourcePath = referenceSourcePath;
+    return this;
+  }
+
+  /**
+   * Read reads from the given path. The input files may be in any format (BAM/CRAM/SAM).
+   *
+   * @param path the file or directory to read from
+   * @return a {@link HtsjdkReadsRdd} that allows access to the reads
+   * @throws IOException if an IO error occurs while determining the format of the files and reading
+   *     the header
+   */
+  public HtsjdkReadsRdd read(String path) throws IOException {
+    return read(path, null);
+  }
+
+  /**
+   * Read reads from the given path, using the given traversal parameters to filter the reads. The
+   * input files may be in any format (BAM/CRAM/SAM).
+   *
+   * @param path the file or directory to read from
+   * @param traversalParameters parameters that determine which reads should be returned, allows
+   *     filtering by interval
+   * @param <T> the type of Locatable for specifying intervals
+   * @return a {@link HtsjdkReadsRdd} that allows access to the reads
+   * @throws IOException if an IO error occurs while determining the format of the files
+   */
+  public <T extends Locatable> HtsjdkReadsRdd read(
+      String path, HtsjdkReadsTraversalParameters<T> traversalParameters) throws IOException {
+
+    FileSystemWrapper fileSystemWrapper =
+        useNio ? new NioFileSystemWrapper() : new HadoopFileSystemWrapper();
+
+    String firstSamPath;
+    if (fileSystemWrapper.isDirectory(sparkContext.hadoopConfiguration(), path)) {
+      firstSamPath =
+          fileSystemWrapper.firstFileInDirectory(sparkContext.hadoopConfiguration(), path);
+    } else {
+      firstSamPath = path;
+    }
+    SamFormat samFormat = SamFormat.fromPath(firstSamPath);
+
+    if (samFormat == null) {
+      throw new IllegalArgumentException("Cannot find format extension for " + path);
+    }
+
+    AbstractSamSource abstractSamSource;
+    switch (samFormat) {
+      case BAM:
```

Comment:
Could this enums have a method to create the `AbstractSamSource` instead of this switch/case statements?

Reply:
Moved to the `SamFormat` enum.